### PR TITLE
build: move packaging in separate folder

### PR DESCRIPTION
**Pull request by DevVersion**
Moves all related packaging code into a `packaging` folder inside of gulp.

At some point we can move the folder completely to `tools/` and can share it with other repositories like `flex-layout`.

### CODE REVIEW

File: tools/gulp/packaging/build-utils.ts [hunk 1]
```diff
@@ -0,0 +1,85 @@
+import {writeFileSync, readFileSync} from 'fs-extra';
+import {join} from 'path';
+import {LICENSE_BANNER, MATERIAL_VERSION} from '../constants';
```
Comment (jelbourn) [inline tools/gulp/packaging/build-utils.ts]: I really dislike grab-bag "util" files. Can we break this up into individual files?

File: tools/gulp/packaging/build-functions.ts [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+  copyFiles(sourcePath, 'package.json', releasePath);
+
+  updatePackageVersion(releasePath);
+  createTypingFile(releasePath, packageName);
+  createMetadataFile(releasePath, packageName);
+  addPureAnnotationsToFile(join(releasePath, '@angular', `${packageName}.es5.js`));
+}
+
+/** Builds the bundles for the specified package. */
+export async function buildPackageBundles(entryFile: string, packageName: string) {
```
Comment (jelbourn) [inline tools/gulp/packaging/build-functions.ts]: Maybe put `composeRelease` and `buildPackageBundles` into their own files?

File: tools/gulp/packaging/inline-resources.ts [hunk 3]
```diff
@@ -32,3 +34,24 @@ export function inlineMetadataResources(metadata: any, componentResources: Map<s
     }
   }
 }
+
+
+/** Inlines HTML and CSS resources into `metadata.json` files. */
+export function inlinePackageMetadataFiles(packagePath: string) {
```
Comment (jelbourn) [inline tools/gulp/packaging/inline-resources.ts]: I'd prefer this to stay in its own file

File: tools/gulp/packaging/ts-compiler.ts [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+  fs.writeFileSync(outputPath, transpiled.outputText);
+
+  if (transpiled.sourceMapText) {
+    fs.writeFileSync(`${outputPath}.map`, transpiled.sourceMapText);
+  }
+}
+
+/** Formats the TypeScript diagnostics into a error string. */
+export function formatDiagnostics(diagnostics: ts.Diagnostic[], baseDir: string): string {
```
Comment (jelbourn) [inline tools/gulp/packaging/ts-compiler.ts]: Separate transpile and diagnostics into separate files?

Comment (DevVersion) [inline tools/gulp/packaging/ts-compiler.ts]: Ugh this ends up in a lot of files. If we do this then we should have a folder `packaging/typescript`?

File: tools/gulp/packaging/build-utils.ts (hunk 1 above)
Comment (DevVersion) [inline tools/gulp/packaging/build-utils.ts]: Not sure I'm a bit afraid of having 5 files just for small functions. What do you think?

File: tools/gulp/packaging/build-functions.ts (hunk 2 above)
Comment (DevVersion) [inline tools/gulp/packaging/build-functions.ts]: I like that idea. Do you have any good ideas for the file names?

- `build-release.ts`
- `build-normal.ts`

Something like that?

File: tools/gulp/packaging/build-utils.ts (hunk 1 above)
Comment (jelbourn) [inline tools/gulp/packaging/build-utils.ts]: Small, single-purpose files are just much easier to grok than have a bunch of unrelated stuff in one file.

File: tools/gulp/packaging/ts-compiler.ts (hunk 4 above)
Comment (jelbourn) [inline tools/gulp/packaging/ts-compiler.ts]: I don't think it would need a a separate directory- maybe just `transpile.ts` and `typescript-diagnostics.ts`?

File: tools/gulp/packaging/build-functions.ts (hunk 2 above)
Comment (jelbourn) [inline tools/gulp/packaging/build-functions.ts]: Why not `build-bundles.ts` and `compose-release.ts`?

Comment (DevVersion) [inline tools/gulp/packaging/build-functions.ts]: Sounds good. I like that

File: tools/gulp/packaging/ts-compiler.ts (hunk 4 above)
Comment (DevVersion) [inline tools/gulp/packaging/ts-compiler.ts]: Sounds a bit unnecessary to be honest. The diagnostics are actually part of the "custom" ts-compiler.

File: tools/gulp/packaging/build-utils.ts (hunk 1 above)
Comment (DevVersion) [inline tools/gulp/packaging/build-utils.ts]: Okay I will try to see how I can split them up.

File: tools/gulp/packaging/ts-compiler.ts (hunk 4 above)
Comment (jelbourn) [inline tools/gulp/packaging/ts-compiler.ts]: If they're not used outside this file then they can stay there without being exported.

Comment (DevVersion) [inline tools/gulp/packaging/ts-compiler.ts]: Yeah good point with the `export`.

**Comment by DevVersion** [comment]
@jelbourn Moved the different build utils into separate files. Let me know if you want any further changes!
